cocotb/bfmgen.py

```diff
@@ -0,0 +1,364 @@
+###############################################################################
+# Copyright cocotb contributors
+# Licensed under the Revised BSD License, see LICENSE for details.
+# SPDX-License-Identifier: BSD-3-Clause
+###############################################################################
+import os
+import argparse
+import importlib
+from string import Template
+
+import cocotb
+from cocotb.bfms import BfmMgr
+
+
+def bfm_load_modules(module_l): 
+    for m in module_l:
+        try:
+            importlib.import_module(m)
+        except Exception as e:
+            print("Error: failed to load module \"" + str(m) + "\": " + str(e))
+            raise e
+
+def process_template_vl(template, info):
+    t = Template(template)
+    
+    bfm_import_calls = ""
+    for i in range(len(info.import_info)):
+        imp = info.import_info[i]
+        bfm_import_calls += "              " + str(i) + ": begin\n"
+        bfm_import_calls += "                  " + imp.T.__name__  + "(\n"
+        for pi in range(len(imp.signature)):
+            p = imp.signature[pi]
+            if p.ptype.s:
+                bfm_import_calls += "                      $cocotb_bfm_get_param_i32(bfm_id)"
+            else:
+                bfm_import_calls += "                      $cocotb_bfm_get_param_ui32(bfm_id)"
+            
+            if pi+1 < len(imp.signature):
+                bfm_import_calls += ","
+            bfm_import_calls += "\n"
+        bfm_import_calls += "                      );\n"
+        bfm_import_calls += "              end\n"
+        
+    bfm_export_tasks = ""
+    for i in range(len(info.export_info)):
+        exp = info.export_info[i]
+        bfm_export_tasks += "    task " + exp.T.__name__ 
+        
+        if len(exp.signature) > 0:
+            bfm_export_tasks += "("
+            for j in range(len(exp.signature)):
+                p = exp.signature[j]
+                bfm_export_tasks += p.ptype.vl_type() + " " + p.pname
+                if j+1 < len(exp.signature):
+                    bfm_export_tasks += ", "
+            bfm_export_tasks += ");\n"
+        else:
+            bfm_export_tasks += ";\n"
+        bfm_export_tasks += "    begin\n"
+        bfm_export_tasks += "        $cocotb_bfm_begin_msg(bfm_id, " + str(i) + ");\n"
+        for p in exp.signature:
+            if p.ptype.s:
+                bfm_export_tasks += "        $cocotb_bfm_add_param_si(bfm_id, " + p.pname + ");\n"
+            else:
+                bfm_export_tasks += "        $cocotb_bfm_add_param_ui(bfm_id, " + p.pname + ");\n"
+            
+        bfm_export_tasks += "        $cocotb_bfm_end_msg(bfm_id);\n"
+        bfm_export_tasks += "    end\n"
+        bfm_export_tasks += "    endtask\n"
+        
+    
+    impl_param_m = {
+        "bfm_classname" : info.T.__module__ + "." + info.T.__qualname__,
+        "bfm_import_calls" : bfm_import_calls,
+        "bfm_export_tasks" : bfm_export_tasks
```

Comment:
nit: PEP8 wants no whitespace before the `:`:
```suggestion
        "bfm_export_tasks": bfm_export_tasks
```

In practice, I like to write this type of thing using `dict` so that the names are unquoted:
```
impl_param_m = dict(
    bfm_classname=info.T.__module__ + "." + info.T.__qualname__,
    bfm_import_calls=bfm_import_calls,
    bfm_export_tasks=bfm_export_tasks
)
```

Reply:
Moved to using dict()